### ISSUE TITLE
Reference type parameters product invalid IL code when patch method parameter is not reference type

### ISSUE THREAD

**Issue by aviadmini** [pardeike/Harmony#79]
Patching Unity game (Unturned), method
```csharp
private void doDamage(byte amount, Vector3 newRagdoll, EDeathCause newCause, ELimb newLimb, CSteamID newKiller, out EPlayerKill kill)
```
Prefix
```csharp
static bool PlayerDamagePrefix(PlayerLife __instance, ref byte amount, ref Vector3 newRagdoll, ref EDeathCause newCause, ref ELimb newLimb, ref CSteamID newKiller, out EPlayerKill kill)
```
Postfix
```csharp
static void PlayerDamagePostfix(PlayerLife __instance, byte amount, Vector3 newRagdoll, EDeathCause newCause, ELimb newLimb, CSteamID newKiller, ref EPlayerKill kill)
```
This produces IL code if `ref` is removed from `ref EPlayerKill kill` parameter in postfix:
```
System.InvalidProgramException: Invalid IL code in (wrapper dynamic-method) SDG.Unturned.PlayerLife:doDamage_Patch2 (object,byte,UnityEngine.Vector3,SDG.Unturned.EDeathCause,SDG.Unturned.ELimb,Steamworks.CSteamID,SDG.Unturned.EPlayerKill&): IL_0fd6: call      0x00000415
```

I believe this behavior is logical but it was pain to figure out

**Comment by pardeike** [comment]
That's a simple side effect from the call structure of the dynamic replacement method. It has the same signature of the original method and calls all prefixes, then the original and then all postfixes. C# requires that you assign an out-var before you potentially use it or before you potentially return it. Since Postfixes are always called, but Prefixes are potentially skipped if a previous Prefix returns false, neither your Prefix nor the the Original is called which means you cannot pass the parameter by ref into the Postfix because that would require it to be initialized first.

Btw: turning debugging on with `HarmonyInstance.DEBUG = true` will give you an instant overview of where the error appears and that would have been when the patch calls your postfix.

I'm going to update the wiki to make this clearer.

**Comment by pardeike** [comment]
I just double checked. Harmony supports either `ref` or `out` in either Prefix or Postfix patches. Must be something else. Can I get your harmony debug output for that method and a description of the parameter types (struct, enum, class, and which struct size) of the original method?

**Comment by aviadmini** [comment]
The exception is thrown on Harmony.Patch call, not when postfix is called. And there is no exception when postfix has `ref` paremeter

**Comment by pardeike** [comment]
I replicated your example in a minimal way:
```csharp
using Harmony;
using System;

public class TestClass
{
	public static void Test(out string foo)
	{
		foo = "foo";
	}
}

[HarmonyPatch(typeof(TestClass), "Test")]
public class Patch1
{
	public static bool Prefix(out string foo)
	{
		foo = "bar";
		Console.WriteLine("Prefix1: " + (foo ?? "null"));
		return false;
	}

	public static void Postfix(string foo)
	{
		foo = "bar2";
		Console.WriteLine("Postfix1: " + (foo ?? "null"));
	}
}

public class Test
{
	static void Main(string[] args)
	{
		var harmony = HarmonyInstance.Create("Test");
		harmony.PatchAll();

		TestClass.Test(out var c);
		Console.WriteLine(c);
		Console.WriteLine("Done");
		Console.ReadKey();
	}
}
```

**Comment by pardeike** [comment]
Adding `ref` or `out` to the Postfix works as well and the result changes from `bar` to `bar2`.

**Comment by pardeike** [comment]
> I'm using the 0Harmony.dll from Releases
That makes a big difference. I am actively testing the master branch with Achtung right now and it works well. Maybe you should not yet release your mod with it but you can try and see if your problem is solved in Harmony v1.1

**Comment by pardeike** [comment]
Please test with latest HEAD version (not the 1.1 release) and open this issue if it still fails.
